safenode/src/protocol/online_transfers/mod.rs

```diff
@@ -0,0 +1,389 @@
+    if total_output_amount > total_input_amount {
+        return Err(Error::NotEnoughBalance(total_input_amount.to_string()));
+    }
+    Ok(())
+}
+
+/// The tokens of the input dbcs will be transfered to the
+/// new dbcs (and a change dbc if any), which are returned from this function.
+/// This does not register the transaction in the network.
+/// To do that, the `signed_spends` of each new dbc, has to be uploaded
+/// to the network. When those same signed spends can be retrieved from
+/// enough peers in the network, the transaction will be completed.
+#[allow(clippy::result_large_err)]
+fn crate_transfer_with(selected_inputs: Inputs) -> Result<Outputs> {
+    let Inputs {
+        dbcs_to_spend,
+        recipients,
+        change: (change, change_to),
+    } = selected_inputs;
+
+    let mut inputs = vec![];
+    for (dbc, derived_key) in dbcs_to_spend {
+        let revealed_amount = match dbc.revealed_amount(&derived_key) {
+            Ok(amount) => amount,
+            Err(err) => {
+                warn!("Ignoring dbc, as it didn't have the correct derived key: {err}");
+                continue;
+            }
+        };
+        let input = InputHistory {
+            input: RevealedInput::new(derived_key, revealed_amount),
+            input_src_tx: dbc.src_tx,
+        };
+        inputs.push(input);
+    }
+
+    let mut tx_builder = TransactionBuilder::default()
+        .add_inputs(inputs)
+        .add_outputs(recipients);
+
+    let mut rng = rng::thread_rng();
+
+    let dbc_id_src = change_to.random_dbc_id_src(&mut rng);
+    let change_id = dbc_id_src.dbc_id();
+    if change.as_nano() > 0 {
+        tx_builder = tx_builder.add_output(change, dbc_id_src);
+    }
+
+    // Finalize the tx builder to get the dbc builder.
+    let dbc_builder = tx_builder
+        .build(Hash::default(), &mut rng)
+        .map_err(Error::Dbcs)?;
+
+    // Perform validations of input tx and signed spends,
+    // as well as building the output dbcs.
+    let mut created_dbcs: Vec<_> = dbc_builder
+        .build()
+        .map_err(Error::Dbcs)?
+        .into_iter()
+        .map(|(dbc, amount)| CreatedDbc { dbc, amount })
+        .collect();
+
+    let mut change_dbc = None;
+    created_dbcs.retain(|created| {
+        if created.dbc.id() == change_id {
+            change_dbc = Some(created.dbc.clone());
+            false
+        } else {
+            true
+        }
+    });
+
+    Ok(Outputs {
+        created_dbcs,
+        change_dbc,
+    })
+}
+
+async fn get_fees(dbc_id: DbcId, client: &Client) -> Result<BTreeMap<NodeId, RequiredFee>> {
+    let request = Request::Query(Query::Spend(SpendQuery::GetFees {
+        dbc_id,
+        priority: SpendPriority::Normal,
+    }));
+    let responses = client
+        .send_to_closest(request)
+        .await
+        .map_err(|e| Error::CouldNotGetFees(e.to_string()))?;
+
+    // We just want to receive at least a majority of results, we don't care about any errors
```

Comment:
shall an error to be raised in case of just received less than majority valid fees?

Reply:
That is handled in thae calling code above. And we don't need to raise an error for that. We just skip that input dbcs and try with the next.

If we in the end didn't have enough tokens in the successful dbcs, _then_  we error.

Reply:
yeah, I understand the `caller` can carry out the mojority check.
However, if the `majority` is always required, then it will be more safe/convenient to make the check included in this `get` func, so that no extra work need to be undertaken outside it?

Reply:
That is actually not more convenient. And there is no extra work.

We are using the `required_responses` variable for additional checks further down after ln 148, see line 165 to ln 170.

```
let max_invalid_fees = num_responses - required_responses;
            if invalid_fees.len() > max_invalid_fees {
                let valid_responses = num_responses - invalid_fees.len();
                warn!("Not enough valid fees received from nodes to spend the input. Found: {valid_responses}, needed: {required_responses}", );
                continue;
            }
```

Reply:
Get fees gets what it can get. Caller decides if it is enough or not.

It makes more sense to keep that type of validation (whether enough or not) on the one and same level, instead of sprinkling it out over several different functions.